sql/2020/07_SEO/json-ld-scripts-raw.sql

```diff
@@ -0,0 +1,98 @@
+# how many lson-ld scripts on a page
+SELECT 
+ARRAY_LENGTH(REGEXP_EXTRACT_ALL(body, '(?i)<script[^>]*type=[\'"]?application\\/ld\\+json[\'"]?[^>]*>(.*?)<\\/script>')) AS jsonld_scripts_count,
```

Comment:
Proposed custom metric:

```js
document.querySelectorAll('script[type="application/ld+json" i]').length
```

Reply:
This one probably needs to be on the raw body as it's designed to find out if the tags parses into json. But thinking about it, I think I can read the raw text of a script in the DOM and try to parse it there. Will test.

Reply:
What does the i in the selector signify?

Reply:
Case insensitivity. Plug the reflex in here: https://regexr.com/ and make sure you choose PCRE instead of JavaScript and it should explain each part to you.

Reply:
Sorry, talking about the css selector from @rviscomi . I though that, but tested and it's case insensitive without it.

Reply:
Oh right sorry. It’s case insensitivity too: https://stackoverflow.com/questions/38399089/ignoring-case-sensitiveness-in-queryselectorall/38399344#38399344

Not sure why your testing showed this wasn’t necessary though. Classes are [somewhat case insensitive](https://stackoverflow.com/questions/12533926/are-class-names-in-css-selectors-case-sensitive) by default but attribute values are usually not.

Reply:
I just did this and it still found all the scripts that are lower case 🤷‍♂️

`document.querySelectorAll('Script[Type="APplication/ld+json"]')`

Reply:
I just discovered that some attributes are case sensitive, so my gathering for meta tags is missing a whole bunch of data. I opened an issue for it:

https://github.com/HTTPArchive/legacy.httparchive.org/issues/188

What's the best way to resolve this. Do I need to fork and fix to get it into the next crawl?

Reply:
It'd be nice to fix for the future but it won't be included until the October crawl at the soonest, the results of which would be available too late to be included in the Almanac. I don't think the case sensitivity would significantly affect the results, but if that's a concern we can include a side note about it in the chapter.

Reply:
I mentioned to the team that we might be able to do a raw html crawl at the end to get anything we missed. If not, the side note will work.

My concern is that older websites tended to shout (upper case) their html, so they would be the ones we miss.

Reply:
Can we submit the PR to fix this now while it’s fresh in our mind? Really want to avoid what we did last year which was not think about custom metrics for 11 months and then have a complete blind panic for the last month! Would be really nice to work on issues like this (or even custom metrics we wish we’d added in hindsight!) as we analyse the 2020 results so we’ve a head start on 2021.

Would also mean you can have a quick look during (or at end of) October and if the stats are radically different then can decide what to do.

Fix looks relatively simple since we’re using Chrome for the crawl: https://stackoverflow.com/questions/38399089/ignoring-case-sensitiveness-in-queryselectorall

Reply:
Will do